### ISSUE TITLE
Investigate the performance impact of CPU/GPU data copying for livepeer_dnn filter

### ISSUE THREAD

**Issue by yondonfu** [livepeer/lpms#250]
We suspect that CPU/GPU data copying in the lvpdnn filter is a performance bottleneck in the transcoding pipeline causing transcoding to slow down substantially when there is a large amount of streams being transcoded. 

We should be able to test whether this is a bottleneck by simulating a ton of streams coming into a local LPMS setup (or maybe using `livepeer_bench`) and monitoring transcoding speed.

**Comment by cyberj0g** [comment]
I created a pretty comprehensive benchmarking script for `livepeer_bench`, and wasn't yet able neither prove nor dismiss hypothesis that PCI-E x1 limits transcoding performance because of CPU<>GPU memory copying in `livepeer_dnn` filter. The reason for that is benchmarking process hitting other bottlenecks, before getting to (indirectly) test mem copy bottleneck hypothesis.

**Test rig specs:** 2 x GTX 1070 (8 Gb) - only GPU 0 is used; 8 Gb RAM; CPU i7-6700K with 7 cores available.

**Test process**: sequentially invoking `livepeer_bench` with all reasonable combinations of `detectionFrequency` and `concurrentSessions` parameters, while measuring CPU and RAM usage by polling `top`, and GPU usage using NVIDIA Management Library. [Script](https://github.com/cyberj0g/livepeer-misc/blob/main/ai/detector_bench.py) and [obtained data](https://github.com/cyberj0g/livepeer-misc/blob/main/ai/bench_results.csv). [Notebook](https://github.com/cyberj0g/livepeer-misc/blob/main/ai/bench_analysis.ipynb) with analysis.

**Hypothesis test**: with this process, PCI-E overload should be indirectly indicated by failed runs (with 'fail' determined by live transcoding requirements), while system load metrics are below 100%.

**Findings**:

1. All failed runs are caused by either RAM or VRAM exhaustion. No runs are failed because of suspected PCI-E bus overload. It can be seen by red markers locations on dots - none of red markers are within normal memory usage for this test system.
2. Very high RAM usage by `livepeer_dnn` filter should be investigated and addressed before looking into CPU<>GPU memory copying issue.
3. It may be worth looking into more direct way of measuring PCI-E performance. NVIDIA API might provide [such way](https://www.tomshardware.com/reviews/graphics-performance-myths-debunked,3739-4.html).

**Comment by jailuthra** [comment]
Great work @cyberj0g - I really dig the script and python notebook used for analysis!

> All failed runs are caused by either RAM or VRAM exhaustion. 

Hmm makes sense that we're running out of VRAM before PCIe lane is becoming a bottleneck, we were using ~2.3GB less VRAM on the signature filter tests before, where we saw PCIe bottlneck.

I think we should give it another shot by  changing `-detectionSampleRate` as initially proposed by @yondonfu [here](https://github.com/livepeer/lpms/issues/250#issuecomment-907452230) and hardcoding `-detectionFrequency` to `1` - instead of the other way round like done right now.

Overall amount of data being copied b/w CPU <> GPU should still scale similarly - but with `detectionFrequency` we're deciding to run or not run the filter at segment-level, and with `detectionSampleRate` we decide it on the frame-level.

Right now with a `detectionFrequency > 2` we might not see a lot of overlap between different sessions running DNN filter simultaneously, as there is less likelihood that two different sessions overlap with using the DNN filter, and thus the CPU <> GPU copying may be happening in big discrete and separate chunks, as opposed to together for all the concurrent sessions.

> Very high RAM usage by livepeer_dnn filter should be investigated and addressed before looking into CPU<>GPU memory copying issue.

Good catch - RAM usage somehow went totally under the radar while developing this. Makes sense to investigate why Tensorflow is eating up so much RAM for each new session - and if the older model loading API we had can help with that cc @oscar-davids

**Comment by cyberj0g** [comment]
Ran tests with `detectionSampleRate` parameter, @jailuthra thanks for pointing this out. Unless there's some mistake, results are very similar, I suspect there is a memory leak somewhere in the filter, and RAM usage depends on number of frames processed, regardless of how they are picked. So it seems CPU<>GPU bottleneck still can't be tested. I will try to profile the filter to figure out what eats up the RAM.

As of measuring PCI-E bandwidth usage directly, looks like the easiest approach is to use [Intel PCM API](https://github.com/opcm/pcm). The catch is, bus bandwidth counter requires server CPU, e.g. Intel Xeon. Does anybody have access to such rig? Having PCI-E x1 is not mandatory, because we'll measure absolute data transfer speed and will know if it exceeds 1x.

**Comment by jailuthra** [comment]
Thanks for running the tests again @cyberj0g!

>  I suspect there is a memory leak somewhere in the filter, and RAM usage depends on number of frames processed, regardless of how they are picked

Ah really good catch! I think I might have found it after going through my recent PR for that filter https://github.com/livepeer/FFmpeg/pull/9/files#diff-aa75e2ee0617a32a0641181edd48853f81052680fc6ae875d1be52142103e856R226
The `AVFrame *out` is allocated but never freed if the DNN model runs successfully. I will send a quick PR for this by tonight!

And this reminds me to set up memory profiling in our end-to-end testing before any major changes on the C side... happy to discuss the best way to go about this with you sometime :)

> The catch is, bus bandwidth counter requires server CPU, e.g. Intel Xeon.

Hmm maybe @iameli can help here

**Comment by oscar-davids** [comment]
Great work @cyberj0g!
I found that the current CPU occupies 100% is when the CPU uploads the model to the GPU when loading the model. This seems to be the PCI-E bandwidth problem. I think it is effective to control the detectionSampleRate parameter except for this time.
I think it seems most effective for us to load the model once at the start of the app.

**Comment by cyberj0g** [comment]
Memory leak should be [fixed](https://github.com/livepeer/FFmpeg/pull/12) now, will run tests again shortly.

**Comment by cyberj0g** [comment]
Interesting, it seems we still cannot approach potential PCI-E issue, while running detection model with TF backend, because now VRAM is a bottleneck - it passes transcoding speed test, before failing with CUDA OOM at about 12 sessions.

**Comment by yondonfu** [comment]
Looks like the VRAM bottleneck with detection results in a capacity decrease of 20 - 12 = 8 sessions which represents a 40% decrease in capacity on a 1080.

Based a conversation with @cyberj0g, we discussed the following options to address this bottleneck:

**Maximize TF VRAM Sharing**

Each transcoding session is associated with a TF session right now. We know that a bunch of VRAM is shared between TF sessions already, but there is still additional VRAM usage per TF session. We could try to maximize TF VRAM sharing by eliminating the additional VRAM usage per TF session. One way to do this might be to use a single TF session across all transcoding sessions (i.e. by passing a pointer to the same TF session to each transcoding session). The unknown here is thread safety.

Previously, @jailuthra shared some VRAM usage data in [this Discord comment](https://discord.com/channels/423160867534929930/694670250901504100/874405265527689226):

- 140 mb for a session without detection enabled
- 2309 mb for creating the first session with detection enabled
- 250 mb for creating an additional session with detection enabled
- 250 - 140 = 110 mb additional VRAM usage per additional session with detection enabled relative to a session without detection enabled

These numbers may not be completely accurately depending on the environment, but could be useful as a reference...

Presumably, the 2309 mb for creating the first session with detection enabled is VRAM that is already shared across sessions. So, an update that shares all VRAM across all sessions would only eliminate the additional 110 mb VRAM usage per additional session with detection enabled but, in theory, this would roughly double the available VRAM available.

Something to note here is that we've typically observed NVDEC utilization to be the bottleneck on GPUs such as the 1080 where transcoding speed slows down to slower than real-time due to there only being a single NVDEC chip before all VRAM on the GPU is consumed. So, it is surprising that with detection enabled, we exhaust all VRAM. If we use Jai's numbers from above, there would be 8000 mb - 2309 mb = 5691 mb VRAM available after creating the first session with detection and if each additional session consumes 250 mb then there would be room for ~22 sessions (5691 / 250) before exhausting VRAM. But Ivan's results show VRAM exhausted at 12 sessions. Could be the case that these numbers are a bit off.  

**Backend Alternatives**

We could use a backend that is less VRAM hungry than TF. Alternatives include PyTorch, ONNX. Or we could look into using something like [Triton](https://github.com/triton-inference-server/server) which in theory would support multiple backends.

**Running Model on CPU**

The model is currently using the mobilenetv2 architecture so it should be pretty lightweight to run on a CPU. In order to avoid a PCIe bottleneck for downloading decoded frames, the frames could be scaled down on the CUDA cores before downloading to CPU since the model expects low resolution images anyway. This would be similar to what we're planning on doing for the [signature filter](https://github.com/livepeer/FFmpeg/issues/11#issuecomment-916315589). The risk here is that the CPU might become the bottleneck even if the model is lightweight to run since CPU utilization will increase.

As the immediate next step, we're thinking of looking into maximizing TF VRAM sharing first before the other two options.

cc @jailuthra @oscar-davids for thoughts as well

**Comment by cyberj0g** [comment]
@yondonfu thanks for great summary. I implemented approach when both the model and TF session are shared across all filter graph instances. It seem to work well, I can run much more sessions and getting almost constant 100% GPU load. The implementation may need some cross-platform thread synchronization code, in case production use case of `ffmpeg.Transcoder` differs from how `livepeer_bench.go` works, I didn't yet investigate it. In case you want to check the code, see [dnn-benchmark](https://github.com/livepeer/FFmpeg/tree/dnn-benchmark) branch. I will run the benchmark and create PR with more details shortly.

**Comment by cyberj0g** [comment]
Unfortunately, after running full test I couldn't confirm any speed up or from DNNContext (TF model and session) sharing. Looks like Tensorflow uses some caching mechanism internally and won't actually load GraphDef with the same name again, even if it's a new TF Session. Session doesn't seem to directly control GPU resource allocation. The fact sessions are gone in TF 2.0 is indirect proof of that. However, there is some effect on RAM usage. Each new model initialization consumes about 50 Mb RAM. As of performance, it seem to be some minor thread synchronization slowdown (inside TF) when sharing the context. See the table.
```
/livepeer_bench  -in ../../data/bbb/source.m3u8 -transcodingOptions P240p30fps16x9 -concurrentSessions 15 -detectionFreq 1 -detectionSampleRate 1 -nvidia 0 -outPrefix /tmp/
```

| Metric | With sharing | W/O sharing
| --- | ----------- | ----------- |
| RAM, gb | 5.8 | 6.6 |
| VRAM, gb | 7.458 | 7.458 |
| CPU, % | 380 | 380 |
| GPU, % | 95 | 95 |
| RT Duration Ratio (3 runs avg) | 5.164  | 4.976 |

On my GTX 1070M with 8 Gb VRAM, max number of sessions is 15. Model initialization takes 2251 Mb of VRAM, then there's no visible VRAM overhead from TF for each new transcoding session, with or without DNNContext sharing. NVENC takes about 347 MB VRAM per session. Same test on the server with GTX 1070 Ti and PCI-E 1x shows much lower RT duration ratio of about 3.7, which is consistent with [performance difference between these GPUs](https://gpu.userbenchmark.com/Compare/Nvidia-GTX-1070-Ti-vs-Nvidia-GTX-1070-Mobile/3943vsm164129) - no evidence of PCI-E bandwidth bottleneck again.

**TLDR**: no positive effect from sharing TF objects between transcoding sessions, PCI-E bottleneck still not confirmed. Next plan is to explore CPU mode of TF backend.

**Comment by cyberj0g** [comment]
After all tests, I think it's safe to assume that PCI-E bottleneck is not an issue with current detection model and Tensorflow backend, because either max VRAM or max GPU load will be reached first, at least on consumer-grade GPU. [Notebook](https://github.com/cyberj0g/livepeer-misc/blob/main/ai/bench_analysis.ipynb) updated again - more accurate data this time, but not much in terms of new insights (max number of sessions there is lower than in comment above, because of using standard 4 rendition types instead of 1).

**Comment by yondonfu** [comment]
Thanks for the update @cyberj0g. Closing this issue for now since we've completed the current investigation of CPU/GPU data copying impact and have included that VRAM/GPU load becomes a bottleneck before PCIe bandwidth becomes a bottleneck.
